cpp/src/arrow/io/memory.cc

```diff
@@ -110,6 +110,21 @@ Status BufferOutputStream::Write(const void* data, int64_t nbytes) {
   return Status::OK();
 }
 
+Status BufferOutputStream::Seek(int64_t position) {
```

Comment:
I'd prefer creating a separate class for the seekable output stream as its semantics is append-only and downstream applications are using it with that assumption.

Reply:
Do you have recommended class name for seekable output stream?

Reply:
Before discussing the name, could you please justify the use case? 

Also cc @wjones127

Reply:
On the read side we have:

InputStream - non-seekable input
ReadableFile - seekable input

So I think we should use `WriteableFile` and not `OutputStream`.  Also note that some filesystems do not support seekable output (e.g. S3).

Reply:
I optimize my code to use `FixedSizeBufferWriter`, so I don't need this feature now, thanks for your kindly review!